File: backend/src/core/rng.ts
/**
 * Seedable PRNG (mulberry32). Deterministic given a seed, fast enough for
 * gameplay. NOT cryptographic.
 */
export class Rng {
  private state: number;

  constructor(seed: number) {
    // Normalize to unsigned 32-bit.
    this.state = seed >>> 0;
    if (this.state === 0) this.state = 0x9e3779b9; // golden-ratio fallback
  }

  /** Float in [0, 1). */
  nextFloat(): number {
    let t = (this.state += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  /** Integer in [0, max). */
  nextInt(maxExclusive: number): number {
    return Math.floor(this.nextFloat() * maxExclusive);
  }

  /** Uniformly pick an element from a non-empty array. */
  pick<T>(arr: readonly T[]): T {
    if (arr.length === 0) throw new Error("Rng.pick on empty array");
    return arr[this.nextInt(arr.length)]!;
  }

  /** Returns a new shuffled copy of `arr` (Fisher-Yates). */
  shuffle<T>(arr: readonly T[]): T[] {
    const out = arr.slice();
    for (let i = out.length - 1; i > 0; i--) {
      const j = this.nextInt(i + 1);
      [out[i], out[j]] = [out[j]!, out[i]!];
    }
    return out;
  }

  static random(): Rng {
    return new Rng(Math.floor(Math.random() * 0xffffffff));
  }
}

File: backend-rs/src/core/rng.rs
//! Seedable PRNG (mulberry32). Mirrors the TS implementation so that, given
//! the same seed, generated sequences match for deterministic gameplay.

#[derive(Debug, Clone)]
pub struct Rng {
    state: u32,
}

impl Rng {
    pub fn new(seed: u32) -> Self {
        let mut state = seed;
        if state == 0 {
            state = 0x9e37_79b9; // golden-ratio fallback
        }
        Self { state }
    }

    pub fn random() -> Self {
        // Cheap non-deterministic seed sourced from the OS clock. The TS
        // implementation uses Math.random() * 0xffffffff; we use nanos which
        // is good enough for non-cryptographic gameplay.
        let nanos = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.subsec_nanos())
            .unwrap_or(0xdead_beef);
        Self::new(nanos)
    }

    /// Float in [0, 1).
    pub fn next_float(&mut self) -> f64 {
        self.state = self.state.wrapping_add(0x6d2b_79f5);
        let mut t = self.state;
        t = (t ^ (t >> 15)).wrapping_mul(t | 1);
        t ^= t.wrapping_add((t ^ (t >> 7)).wrapping_mul(t | 61));
        let v = (t ^ (t >> 14)) as f64;
        v / 4_294_967_296.0
    }

    /// Integer in [0, max).
    pub fn next_int(&mut self, max_exclusive: u32) -> u32 {
        (self.next_float() * (max_exclusive as f64)) as u32
    }

    /// Uniformly pick an element from a non-empty slice.
    #[allow(dead_code)]
    pub fn pick<'a, T>(&mut self, arr: &'a [T]) -> &'a T {
        assert!(!arr.is_empty(), "Rng.pick on empty array");
        &arr[self.next_int(arr.len() as u32) as usize]
    }

    /// Returns a new shuffled copy (Fisher-Yates).
    #[allow(dead_code)]
    pub fn shuffle<T: Clone>(&mut self, arr: &[T]) -> Vec<T> {
        let mut out = arr.to_vec();
        for i in (1..out.len()).rev() {
            let j = self.next_int((i + 1) as u32) as usize;
            out.swap(i, j);
        }
        out
    }
}
